Mathlib/RingTheory/Extension/Cotangent/Basis.lean

```diff
@@ -0,0 +1,289 @@
+def cotangentEquivProd [Nontrivial S] : D.pres.toExtension.Cotangent ≃ₗ[S]
+    D.presRight.toExtension.Cotangent × S ⊗[D.T] D.presLeft.toExtension.Cotangent :=
+  (D.presLeft.cotangentCompLocalizationAwayEquiv (T := S) D.gbar D.map_ofComp_mk) ≪≫ₗ
+    LinearEquiv.prodComm _ _ _
+
+lemma cotangentEquivProd_symm_apply [Nontrivial S] (x : D.presRight.toExtension.Cotangent)
+      (y : S ⊗[D.T] D.presLeft.toExtension.Cotangent) :
+    D.cotangentEquivProd.symm (x, y) =
+      (D.presLeft.cotangentCompLocalizationAwayEquiv (T := S) D.gbar D.map_ofComp_mk).symm (y, x) :=
+  rfl
+
+/-- The basis of `S ⊗[T] J/J²` induced from the basis on `I/I²`. -/
+def basisLeft : Module.Basis σ S (S ⊗[D.T] D.presLeft.toExtension.Cotangent) :=
+  b.map D.tensorCotangentEquiv.symm
+
+/-- The canonical basis on `(g X - 1)/(g X - 1)²`. -/
+def basisRight : Module.Basis Unit S D.presRight.toExtension.Cotangent :=
+  Generators.basisCotangentAway S D.gbar
+
+/-- The basis on the cotangent space of the constructed presentation. -/
+def basis [Nontrivial S] : Module.Basis (Unit ⊕ σ) S D.pres.toExtension.Cotangent :=
+  (Module.Basis.prod D.basisRight D.basisLeft).map D.cotangentEquivProd.symm
+
+lemma basis_inl [Nontrivial S] :
+    D.basis (.inl ()) =
+      D.cotangentEquivProd.symm (Generators.cMulXSubOneCotangent S D.gbar, 0) := by
+  simpa [basis] using Generators.basisCotangentAway_apply _ _
+
+lemma basis_inr [Nontrivial S] (i : σ) :
+    D.basis (.inr i) = D.cotangentEquivProd.symm (0, D.basisLeft i) := by
+  simp [basis]
+
+lemma pres_val_comp_inr : D.pres.val ∘ Sum.inr = P.val := funext (aeval_X _)
+
+/-- The constructed basis indeed is given by the images of the relations. -/
+lemma basis_apply [Nontrivial S] (r : Unit ⊕ σ) :
+    D.basis r = Extension.Cotangent.mk ⟨D.pres.relation r, D.pres.relation_mem_ker r⟩ := by
+  obtain (r | r) := r
+  · rw [basis_inl, cotangentEquivProd_symm_apply]
+    exact cotangentCompLocalizationAwayEquiv_symm_inr _ _ _
+  · rw [basis_inr, cotangentEquivProd_symm_apply, cotangentCompLocalizationAwayEquiv_symm_inl,
+      basisLeft, Module.Basis.map_apply, tensorCotangentEquiv_symm_apply,
+      LinearMap.liftBaseChange_tmul, one_smul, Extension.Cotangent.map_mk]
+    rfl
+
+end Aux
+
+end
+
+/-- Let `S` be a finitely presented `R`-algebra and suppose `P : R[X] → S` generates `S` with
+kernel `I`. If `I/I²` is free, there exists an `R`-presentation `P'` of `S` extending `P` with
+kernel `I'`, such that `I'/I'²` is free on the images of the relations of `P'`. -/
+@[stacks 07CF]
+lemma exists_presentation_of_free [Algebra.FinitePresentation R S]
+    {α : Type*} (P : Generators R S α) [Finite α] {σ : Type*}
+    (b : Module.Basis σ S P.toExtension.Cotangent) :
+    ∃ (P' : Presentation R S (Unit ⊕ α) (Unit ⊕ σ))
+      (b : Module.Basis (Unit ⊕ σ) S P'.toExtension.Cotangent),
+      P'.val ∘ Sum.inr = P.val ∧
+      ∀ r, b r = Extension.Cotangent.mk ⟨P'.relation r, P'.relation_mem_ker r⟩ := by
```

Comment:
There are two `b`s here. Am I correct that the first `b` does not appear in the goal? Should we just assume `Module.Free S P.toExtension.Cotangent` instead? (and replace `σ` with `Fin (finrank _ _)` or some arbitrary thing with the right cardinality) Nonetheless one of them should be renamed.